sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1548,6 +1548,31 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
         filter
       }
 
+    case filter @ Filter(condition, w: Window) if w.partitionSpec.isEmpty =>
+      w.windowExpressions match {
+        case Seq(alias @ Alias(WindowExpression(_: RowNumber, WindowSpecDefinition(Nil, orderSpec,
+            SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow))), _)) =>
+          val aliasAttr = alias.toAttribute
+          val limitValue = splitConjunctivePredicates(condition).collectFirst {
+            case LessThanOrEqual(e, IntegerLiteral(v)) if e.semanticEquals(aliasAttr) => v
+            case Equality(e, IntegerLiteral(v)) if e.semanticEquals(aliasAttr) => v
+            case LessThan(e, IntegerLiteral(v)) if e.semanticEquals(aliasAttr) => v - 1
+          }
+
+          limitValue match {
+            case Some(lv) if lv <= 0 =>
+              LocalRelation(filter.output, data = Seq.empty, isStreaming = filter.isStreaming)
+            case Some(lv)
+                if lv < conf.topKSortFallbackThreshold && w.child.maxRows.forall(_ > lv) =>
+              filter.copy(child =
+                w.copy(child = Limit(Literal(lv), Sort(orderSpec, true, w.child))))
```

Comment:
One worry is what if other optimizer rules put/move operators between Limit and Sort? Then we can't use the `TakeOrderedAndProjectExec` and introduce a big overhead by this global sort.

Reply:
BTW, is `orderSpec` always specified?

Reply:
Yes. `orderSpec` is always specified for `RowNumber`:
```
spark-sql> SELECT a, b, row_number() OVER (PARTITION BY a) FROM VALUES ('A1', 2), ('A1', 1), ('A2', 3), ('A1', 1) tab(a, b);
Error in query: Window function row_number() requires window to be ordered, please add ORDER BY clause. For example SELECT row_number()(value_expr) OVER (PARTITION BY window_partition ORDER BY window_ordering) from table
spark-sql>
```
But some window function do not need `orderSpec`. For example:
```sql
spark-sql> SELECT a, b, last(b) OVER (PARTITION BY a) FROM VALUES ('A1', 2), ('A1', 1), ('A2', 3), ('A1', 1) tab(a, b);
A2	3	3
A1	2	1
A1	1	1
A1	1	1
```